TreeList Virtual Scrolling (#2089)
telerik/blazor-docs: components/treelist/virtual-scrolling.md

---
title: Virtual Scrolling

> page_title: TreeList - Sorting

Reviewer: ```suggestion
page_title: TreeList - Virtual Scrolling
```

description: Enable and configure the virtual scrolling of rows in the TreeList for Blazor.
slug: treelist-virtual-scrolling
tags: telerik,blazor,treelist,virtual,rows,scrolling
published: True
position: 25
---

# TreeList Virtual Scrolling

> Virtual scrolling provides an alternative to paging. Instead of utilizing a pager, the user vertically scrolls through all records in the data source.

Reviewer: ```suggestion
Virtual scrolling provides an alternative to paging. Instead of utilizing a pager, the user scrolls vertically through all records in the data source.
```

To enhance rendering performance, the TreeList reuses the same set of HTML elements. As the next data loads, a loading indicator appears on the cells. If the user scrolls back up after scrolling down to the next set of rows, the previous data reloads from the data source, similar to regular paging, with the scroll distance determining the data to be loaded.

## Enabling Virtual Scrolling

* Set the `ScrollMode` parameter to `TreeListScrollMode.Virtual` (the default value is `Scrollable`).

> * Set the `Height` and `RowHeight` parameters.

Reviewer: * You may want to give some guidance on how to set these parameters correctly so the developer can reduce the potential issues. You mention them multiple times in the "Notes" section below and they seem pretty important for the proper configuration. 
* You may want to mention how the different values will affect the behavior and performance.

                 Sortable="true"
                 FilterMode="@TreeListFilterMode.FilterMenu">
    <TreeListColumns>
        <TreeListColumn Expandable="true" Field="FirstName" Title="First Name" />
        <TreeListColumn Field="LastName" Title="Last Name" />
        <TreeListColumn Field="Position" />
    </TreeListColumns>
</TelerikTreeList>

@code {
    private List<Employee> TreeListData { get; set; } = new();

    protected override void OnInitialized()
    {
        for (int i = 1; i <= 1000; i++)
        {
            TreeListData.Add(new Employee()
                {
                    Id = i,
                    ParentId = i <= 3 ? null : i % 3 + 1,
                    FirstName = "First " + i,
                    LastName = "Last " + i,
                    Position = i <= 3 ? "Team Lead" : "Software Engineer"
                });
        }

        base.OnInitialized();
    }

    public class Employee
    {
        public int Id { get; set; }
        public int? ParentId { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string Position { get; set; }
    }
}
````

> ## Notes

Reviewer: As discussed, these types of sections aren't user-friendly and don't provide enough context and structure to help the developer with the implementation. Consider transforming this section by using any of the approaches below or a combination of them:
* Split the Notes section into several sections describing specific implementations, their common pitfalls and ways to avoid them.
* Move some of the information to KBs.
* Use sample code with detailed comments explaining what to watch for. If appropriate, use tabs to demonstrate alternative ways.

There are several things to keep in mind when using virtual scrolling:

> * The `RowHeight` is a decimal value that is always considered as pixel values. The TreeList `Height` does not have to be in pixels, but it may help you calculate the `PageSize` (see below).

Reviewer: ```suggestion
* The `RowHeight` is a decimal value that is always interpreted as pixels. The TreeList `Height` does not have to be in pixels, but it may help you calculate the `PageSize` (see below).
```

> * The `RowHeight` is a decimal value that is always considered as pixel values. The TreeList `Height` does not have to be in pixels, but it may help you calculate the `PageSize` (see below).

Reviewer: It is unclear what "see below" refers to.

>     * If the row/cell height the browser would render is larger than the `RowHeight` value, the browser will ignore it. It can depend on the chosen Theme or other CSS rules, or on cell data that falls on more than one row. Inspect the rendered HTML to make sure the grid setting matches the rendering.

Reviewer: * _It_ can depend on—what is "it" referring to?
```suggestion
    * If the row/cell height the browser renders is larger than the `RowHeight` value, the browser will ignore it. It can depend on the chosen Theme, other CSS rules, or cell data that occupies more than one row. Inspect the rendered HTML to make sure the grid setting matches the rendering.
```

        The default TreeList rendering has padding in the cells, and the loading sign has a line height set in order to render. This may impose some minimum heights that can vary with the theme and/or custom styles on the page.

    * The `RowHeight` must not change at runtime, because the new dimensions will cause issues with the scrolling logic.

    * Browser zoom or monitor DPI settings can cause the browser to render different dimensions than the expected and/or non-integer values, which can break the virtualization logic.

* Do not mix virtualization with paging, as they are alternatives to the same feature.

* Provide for a `PageSize` of the TreeList that is large enough, so that the loaded table rows do not fit in the scrollable data area, otherwise the vertical virtual scrollbar will not be created and scrolling will not work. To do this, take into account the `Height` of the TreeList and the `RowHeight`.

>     * The `PageSize` controls how many rows are rendered at any given time, and how many items are requested from the data source when loading data on demand (see below). You should avoid setting large page sizes, you need to only fill up the TreeList data viewport.

Reviewer: I don't see anything below this bullet item. It is unclear what "see below" refers to.

---
title: Virtual Scrolling
page_title: TreeList - Sorting
description: Enable and configure the virtual scrolling of rows in the TreeList for Blazor.
slug: treelist-virtual-scrolling
tags: telerik,blazor,treelist,virtual,rows,scrolling
published: True
position: 25
---

# TreeList Virtual Scrolling

Virtual scrolling provides an alternative to paging. Instead of utilizing a pager, the user vertically scrolls through all records in the data source.

To enhance rendering performance, the TreeList reuses the same set of HTML elements. As the next data loads, a loading indicator appears on the cells. If the user scrolls back up after scrolling down to the next set of rows, the previous data reloads from the data source, similar to regular paging, with the scroll distance determining the data to be loaded.

> ## Enabling Virtual Scrolling

Reviewer 2: "Enabling" seems hard to read. If gerund is a must, consider "Using".

* Set the `ScrollMode` parameter to `TreeListScrollMode.Virtual` (the default value is `Scrollable`).

> * Set the `Height` and `RowHeight` parameters.

Reviewer 2: You can mention explicitly here that one parameter is string, while the other is decimal.

>caption Enable Sorting in Telerik TreeList

````CSHTML
<TelerikTreeList Data="@TreeListData"
                 ScrollMode="@TreeListScrollMode.Virtual"
                 Height="500px"
                 RowHeight="50"
                 IdField="@nameof(Employee.Id)"
                 ParentIdField="@nameof(Employee.ParentId)"
                 Sortable="true"
                 FilterMode="@TreeListFilterMode.FilterMenu">
    <TreeListColumns>
        <TreeListColumn Expandable="true" Field="FirstName" Title="First Name" />
        <TreeListColumn Field="LastName" Title="Last Name" />
        <TreeListColumn Field="Position" />
    </TreeListColumns>
</TelerikTreeList>

@code {

> 

Reviewer 2: ```suggestion
```

>     private List<Employee> TreeListData { get; set; }

Reviewer 2: ```suggestion
    private List<Employee> TreeListData { get; set; } = new();
```

    protected override void OnInitialized()
    {
        TreeListData = new List<Employee>();

        for (int i = 1; i <= 1000; i++)
        {
            TreeListData.Add(new Employee()
                {
                    Id = i,
                    ParentId = i <= 3 ? null : i % 3 + 1,
                    FirstName = "First " + i,
                    LastName = "Last " + i,
                    Position = i <= 3 ? "Team Lead" : "Software Engineer"
                });
        }

        base.OnInitialized();
    }

    public class Employee
    {
        public int Id { get; set; }
        public int? ParentId { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string Position { get; set; }
    }
}
````

> ## Notes

Reviewer 2: Notes is non-structured content that should be avoided. Consider moving everything to another place with better structure. For example, some of the notes can go to the "Using" section and some notes can go to a new "Setting/Using RowHeight" section.

Author: I completely agree that must change the way we present the information in the Notes section. I suggest that we make that decision when the PR for the Grid Virtual Scrolling page is submitted.
